src/transformers/pipelines/document_question_answering.py

```diff
@@ -0,0 +1,449 @@
+
+        return {
+            **encoding,
+            "p_mask": p_mask,
+            "word_ids": word_ids,
+            "words": words,
+        }
+
+    def _forward(self, model_inputs):
+        p_mask = model_inputs.pop("p_mask", None)
+        word_ids = model_inputs.pop("word_ids", None)
+        words = model_inputs.pop("words", None)
+
+        if self.model_type == ModelType.VisionEncoderDecoder:
+            model_outputs = self.model.generate(**model_inputs)
+        else:
+            model_outputs = self.model(**model_inputs)
+
+        model_outputs["p_mask"] = p_mask
+        model_outputs["word_ids"] = word_ids
+        model_outputs["words"] = words
+        model_outputs["attention_mask"] = model_inputs.get("attention_mask", None)
+        return model_outputs
+
+    def postprocess(self, model_outputs, top_k=1, **kwargs):
+        if self.model_type == ModelType.VisionEncoderDecoder:
+            answers = self.postprocess_donut(model_outputs)
+        else:
+            answers = self.postprocess_extractive_qa(model_outputs, top_k=top_k, **kwargs)
+
+        answers = sorted(answers, key=lambda x: x.get("score", 0), reverse=True)[:top_k]
+        if len(answers) == 1:
+            return answers[0]
+        return answers
+
+    def postprocess_donut(self, model_outputs, **kwargs):
+        # postprocess
+        sequence = self.tokenizer.batch_decode(model_outputs.sequences)[0]
+        sequence = sequence.replace(self.tokenizer.eos_token, "").replace(self.tokenizer.pad_token, "")
+        sequence = re.sub(r"<.*?>", "", sequence, count=1).strip()  # remove first task start token
+        ret = {
+            "answer": None,
+        }
+
+        answer = re.search(r"<s_answer>(.*)</s_answer>", sequence)
+        if answer is not None:
+            ret["answer"] = answer.group(1).strip()
+        return [ret]
+
+    def postprocess_extractive_qa(
+        self, model_outputs, top_k=1, handle_impossible_answer=False, max_answer_len=15, **kwargs
+    ):
+        min_null_score = 1000000  # large and positive
+        answers = []
+        words = model_outputs["words"]
+
+        # TODO: Currently, we expect the length of model_outputs to be 1, because we do not stride
+        # in the preprocessor code. When we implement that, we'll either need to handle tensors of size
+        # > 1 or use the ChunkPipeline and handle multiple outputs (each of size = 1).
+        starts, ends, scores, min_null_score = select_starts_ends(
+            model_outputs["start_logits"],
+            model_outputs["end_logits"],
+            model_outputs["p_mask"],
+            model_outputs["attention_mask"].numpy() if model_outputs.get("attention_mask", None) is not None else None,
+            min_null_score,
+            top_k,
+            handle_impossible_answer,
+            max_answer_len,
+        )
+
+        word_ids = model_outputs["word_ids"][0]
+        for s, e, score in zip(starts, ends, scores):
```

Comment:
Can we use better names for s (start) and e (end) here?

Reply:
Yes fixed.